### PR TITLE
gh-99139: Improve NameError error suggestion for instances

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-99139

### CODE REVIEW

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -75,6 +75,27 @@ Important deprecations, removals or restrictions:
 Improved Error Messages
 =======================
 
+* Improve the error suggestion for :exc:`NameError` exceptions for instances.
+  Now if a :exc:`NameError` is raised in a method and the instance has an
+  attribute that's exactly equal to the name in the exception, the suggestion
+  will include ``self.<NAME>`` instead of the closest match in the method
+  scope. Controbuted by Pablo Galindo in :gh:`99139`.
```
Comment (ammaraskar) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  scope. Contributed by Pablo Galindo in :gh:`99139`.
```

File: Python/suggestions.c [hunk 2]
```diff
@@ -226,6 +228,25 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
...
 
+    // Are we inside a method and the instance has an attribute called 'name'?
+    _Py_IDENTIFIER(self);
+    PyObject* self_str = _PyUnicode_FromId(&PyId_self); /* borrowed */
+    if (PySequence_Contains(dir, self_str) > 0) {
+        PyObject* locals = PyFrame_GetLocals(frame);
+        if (!locals) {
+            goto error;
+        }
+        PyObject* self = PyDict_GetItemString(locals, "self"); /* borrowed */
```
Comment (ammaraskar) [inline Python/suggestions.c]:
```suggestion
        PyObject* self = PyDict_GetItem(locals, self_str); /* borrowed */
```

**Review by ammaraskar — APPROVED** [review]
Looks great :)

File: Python/suggestions.c [hunk 3]
```diff
@@ -250,6 +271,10 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
     Py_DECREF(dir);
 
     return suggestions;
+
+error:
+    Py_XDECREF(dir);
```
Comment (isidentical) [inline Python/suggestions.c]: If I am not missing any, all the `goto` sites seem to have `dir` as a `non-NULL` value.
```suggestion
    Py_DECREF(dir);
```

File: Python/suggestions.c [hunk 4]
```diff
@@ -226,6 +228,25 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
...
+            goto error;
+        }
+        PyObject* self = PyDict_GetItemString(locals, "self"); /* borrowed */
+        Py_DECREF(locals);
+        if (!self) {
+            goto error;
+        }
+        
+        if (PyObject_HasAttr(self, name)) {
+            return PyUnicode_FromFormat("self.%S", name);
```
Comment (isidentical) [inline Python/suggestions.c:244]:
```suggestion
            Py_DECREF(dir);
            return PyUnicode_FromFormat("self.%S", name);
```

**Review by isidentical — APPROVED** [review]
Looks really helpful, thanks a lot @pablogsal!

File: Python/suggestions.c [hunk 5]
```diff
@@ -226,6 +228,26 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
         return NULL;
     }
 
+    // Are we inside a method and the instance has an attribute called 'name'?
+    _Py_IDENTIFIER(self);
+    PyObject* self_str = _PyUnicode_FromId(&PyId_self); /* borrowed */
+    if (PySequence_Contains(dir, self_str) > 0) {
```
Comment (kumaraditya303) [inline Python/suggestions.c]: `_PyUnicode_FromId` can return null if allocation fails.  Also we will be deprecating/removing `_Py_IDENTIFIER` as part of https://github.com/python/cpython/issues/90699 so you should use `_Py_ID`.

File: Python/suggestions.c [hunk 6]
```diff
@@ -226,6 +228,26 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
...
+        PyObject* locals = PyFrame_GetLocals(frame);
+        if (!locals) {
+            goto error;
+        }
+        PyObject* self = PyDict_GetItem(locals, self_str); /* borrowed */
```
Comment (kumaraditya303) [inline Python/suggestions.c:239]:
```suggestion
        PyObject* self = PyDict_GetItem(locals, &_Py_ID(self)); /* borrowed */
```

File: Python/suggestions.c (hunk 5 above)
Comment (kumaraditya303) [inline Python/suggestions.c:234]:
```suggestion
    if (PySequence_Contains(dir, &_Py_ID(self)) > 0) {
```

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Python/suggestions.c [hunk 7]
```diff
@@ -226,6 +228,26 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
...
+        Py_DECREF(locals);
```
Comment (kumaraditya303) [inline Python/suggestions.c]:
```suggestion
    if (PySequence_Contains(dir, &_Py_ID(self)) > 0) {
        PyObject* locals = PyFrame_GetLocals(frame);
        if (!locals) {
            goto error;
        }
        PyObject* self = PyDict_GetItem(locals, &_Py_ID(self)); /* borrowed */
        Py_DECREF(locals);
```

File: Python/suggestions.c [hunk 8]
```diff
@@ -1,3 +1,5 @@
+#define NEEDS_PY_IDENTIFIER
+
```
Comment (kumaraditya303) [inline Python/suggestions.c]: This is redundant now.

File: Python/suggestions.c [hunk 9]
```diff
@@ -1,3 +1,5 @@
...
 #include "Python.h"
 #include "pycore_frame.h"
```
Comment (kumaraditya303) [inline Python/suggestions.c:2]:
```suggestion
#include "pycore_frame.h"
#include "pycore_runtime_init.h"  // _Py_ID()
```

Comment (kumaraditya303) [reply Python/suggestions.c:2]: You would also have to regen the global strings.

Comment (pablogsal) [reply Python/suggestions.c:2]: Will do this when I am back at home.

File: Python/suggestions.c [hunk 10]
```diff
@@ -226,6 +227,24 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
         return NULL;
     }
 
+    // Are we inside a method and the instance has an attribute called 'name'?
+    if (PySequence_Contains(dir, &_Py_ID(self)) > 0) {
+        PyObject* locals = PyFrame_GetLocals(frame);
+        if (!locals) {
+            goto error;
+        }
+        PyObject* self = PyDict_GetItem(locals, &_Py_ID(self)); /* borrowed */
```
Comment (serhiy-storchaka) [inline Python/suggestions.c:236]: Do not use `PyDict_GetItem`. It is broken by design.

File: Python/suggestions.c [hunk 11]
```diff
@@ -226,6 +227,24 @@ get_suggestions_for_name_error(PyObject* name, PyFrameObject* frame)
...
+        Py_DECREF(locals);
+        if (!self) {
+            goto error;
+        }
+        
+        if (PyObject_HasAttr(self, name)) {
```
Comment (serhiy-storchaka) [inline Python/suggestions.c:242]: Do not use `PyObject_HasAttr`. It is broken by design.

File: Python/suggestions.c (hunk 10 above)
Comment (pablogsal) [reply Python/suggestions.c:236]: What should we use instead for both?

Comment (serhiy-storchaka) [reply Python/suggestions.c:236]: `PyDict_GetItemWithError` and `PyObject_GetAttr` if limited by public API. It is suggested in the documentation. But `_PyObject_LookupAttr` is more convenient in the latter case.

**Comment by millerdev** [comment]
I'm surprised this uses `hasattr` under the covers. A simple benign name error could cause side effects at run time.
```py
>>> class Test:
...     def __getattr__(self, name):
...         if name == "foo":
...             raise SystemExit("The end.")
...     def bar(self):
...         foo
... 
... Test().bar()
... 
The end.
```

**Pull request merged**